pkg/storage/etcd3/etcdwatcher/watcher.go

```diff
@@ -0,0 +1,340 @@
+	}
+	wc.ctx, wc.cancel = context.WithCancel(ctx)
+	go wc.run()
+	wc.waitEventProc.Add(1)
+	return wc, nil
+}
+
+func (wc *watchChan) run() {
+	go wc.sync()
+	go wc.startWatching()
+	go wc.eventProcessor()
+
+	var errResult *watch.Event
+	select {
+	case err := <-wc.errChan:
+		errResult = wc.parseError(err)
+		wc.cancel()
+	case <-wc.ctx.Done():
+	}
+	// error result is guaranteed to be received by user before closing ResultChan.
+	if errResult != nil {
+		wc.resultChan <- *errResult
+	}
+	wc.waitEventProc.Wait()
+	close(wc.resultChan)
+}
+
+func (wc *watchChan) Stop() {
+	wc.cancel()
+}
+
+func (wc *watchChan) ResultChan() <-chan watch.Event {
+	return wc.resultChan
+}
+
+// sync tries to retrieve existing data send them to process.
+// The revision to watch will be set to the revision in response.
+func (wc *watchChan) sync() {
+	defer close(wc.waitSync)
+	if wc.initialRev != 0 {
+		return
+	}
+
+	opts := []clientv3.OpOption{}
+	if wc.recursive {
+		opts = append(opts, clientv3.WithPrefix())
+	}
+	getResp, err := wc.watcher.client.Get(wc.ctx, wc.key, opts...)
+	if err != nil {
+		go wc.sendError(err)
+		return
+	}
+	wc.initialRev = getResp.Header.Revision
+
+	for _, kv := range getResp.Kvs {
+		wc.sendEvent(parseEtcdKV(kv))
+	}
+}
+
+// startWatching does:
+// - wait for sync to complete
+// - watch on given key and send events to process.
+func (wc *watchChan) startWatching() {
+	select {
+	case <-wc.waitSync:
+	case <-wc.ctx.Done():
+	}
+	opts := []clientv3.OpOption{clientv3.WithRev(wc.initialRev + 1)}
+	if wc.recursive {
+		opts = append(opts, clientv3.WithPrefix())
+	}
+	wch := wc.watcher.client.Watch(wc.ctx, wc.key, opts...)
+	for wres := range wch {
+		if wres.Err() != nil {
+			// If there is an error on server (e.g. compaction), the channel will return it before closed.
+			go wc.sendError(wres.Err())
+			return
+		}
+		for _, e := range wres.Events {
+			wc.sendEvent(parseEtcdEvent(e))
+		}
+	}
+}
+
+// eventProcess does:
+// - receiving event from etcd watcher, process it
+// - queues results if user not receiving fast enough
+// Note: we keep the ordering of results the same as events passed from etcd.
+func (wc *watchChan) eventProcessor() {
+	var toSend watch.Event
+	var outgoing chan watch.Event
+	for {
+		select {
+		case e := <-wc.eventChan:
+			res := wc.transform(e)
+			if res == nil {
+				break
+			}
+			if outgoing == nil {
+				toSend = *res
+				outgoing = wc.resultChan
+			} else {
+				wc.resultQueue.PushBack(*res)
+			}
+		case outgoing <- toSend:
+			if wc.resultQueue.Len() == 0 {
+				outgoing = nil
+				break
+			}
+			elem := wc.resultQueue.Front()
+			wc.resultQueue.Remove(elem)
+			toSend = elem.Value.(watch.Event)
+		case <-wc.ctx.Done():
+			wc.waitEventProc.Done()
+			return
+		}
+	}
+}
+
+// transform transform given event into result for user if not filtered.
+// TODO (Optimization):
+// - Save remote round-trip.
+//   Currently, DELETE and PUT event don't contain the previous value.
```

Comment:
Not sure if this was answered on the other PR but we rely on this for mutations that change labels and filters.  Being able to have the server send that data with it dramatically simplifies the client interaction.